trl/trainer/sft_trainer.py

```diff
@@ -522,33 +521,79 @@ def concat_prompt_completion(example):
                 )
 
                 # Apply the chat template if needed
-                if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
-                    map_kwargs["desc"] = f"Applying chat template to {dataset_name} dataset"
-                column_names = next(iter(dataset)).keys()
-                dataset = dataset.map(
-                    maybe_apply_chat_template,
-                    fn_kwargs={"tokenizer": processing_class},
-                    remove_columns="messages" if "messages" in column_names else None,  # renamed to "text"
-                    **map_kwargs,
-                )
+                first_example = next(iter(dataset))
+                if is_conversational(first_example):
+                    if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
+                        map_kwargs["desc"] = f"Applying chat template to {dataset_name} dataset"
+                    column_names = first_example.keys()
+                    dataset = dataset.map(
+                        apply_chat_template,
+                        fn_kwargs={"tokenizer": processing_class},
+                        remove_columns="messages" if "messages" in column_names else None,  # renamed to "text"
+                        **map_kwargs,
+                    )
+                    add_special_tokens = False  # subsequent tokenization won't add special tokens (mostly for bos)
+                # When dataset is not conversational, we need to add the EOS token at the end of each example
+                # We don't need to do this for conversational datasets as this is already handled by the
+                # `apply_chat_template` function.
+                else:
+                    if isinstance(dataset, Dataset):  # `IterableDataset.map` does not support `desc`
+                        map_kwargs["desc"] = f"Adding EOS to {dataset_name} dataset"
+
+                    def add_eos(example, eos_token):
+                        if "text" in example and not example["text"].endswith(eos_token):  # language modeling case
+                            example["text"] = example["text"] + eos_token
+                        elif "completion" in example and not example["completion"].endswith(eos_token):
+                            example["completion"] = example["completion"] + eos_token
+                        return example
+
+                    dataset = dataset.map(
+                        add_eos,
+                        fn_kwargs={"eos_token": processing_class.eos_token},
+                        remove_columns="messages" if "messages" in column_names else None,  # renamed to "text"
+                        **map_kwargs,
+                    )
+                    add_special_tokens = True  # subsequent tokenization will add special tokens (mostly for bos)
```

Comment:
These changes aren't not related to completion-only loss.
It fixes the case where we add the BOS twice (in `apply_chat_template` and in `tokenize`. To avoid that, we need to pass `add_special_tokens=False` when the chat template was applied.

Reply:
It also fixes the case where we add the EOS twice, because it's isn't found in the end of the sequence, see #3318